### PR TITLE
i18n 

### PR DESCRIPTION
**Pull request by ph1p**
Hi!

This PR adds i18n functionality and is currently WIP. I'll add auto detection and a language chooser or something else.
I have copied some code from buttercup-desktop and experimented a little (:

I am pleased about your thoughts and suggestions.

**[UPDATE]**

- Autodetection

### CODE REVIEW

**Comment by julianpoemp** [comment]
That's what I was looking for 😁👍

**Comment by perry-mitchell** [comment]
Great stuff @ph1p! This'd be super handy.

**Comment by ph1p** [comment]
I'm almost done translating all the strings. Some parts are a bit confusing (`sheets.js`), but it should work (: Currently I have added English and German. Buttercup auto detects the language.

**Comment by perry-mitchell** [comment]
Great @ph1p! Amazing work. Should we give it an in-depth review now? Would you say the WIP is done?

**Comment by ph1p** [comment]
Sure! Feel free to review. I know there's definitely room for improvement, but it's a first version (:

**Comment by ph1p** [comment]
I also tried to fix the build bugs. I made android work, but ios still fails.

File: source/components/ArchivesPage.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ import ToolbarIcon from "./ToolbarIcon.js";
 import { navigateToSearchArchives } from "../actions/navigation.js";
 import { setSearchContext } from "../actions/app.js";
 import { dispatch } from "../store.js";
+import { withTranslation } from "react-i18next";
```
Comment (sallar) [inline source/components/ArchivesPage.js]: Imported but not used

**Comment by ph1p** [comment]
@sallar @perry-mitchell I've also updated the react-navigation, because there is a beta inside the `package.json`. 

**rn-nodeify** > ^10.0.1 (https://github.com/tradle/rn-nodeify/issues/74)
**react-navigation** > ^2.18.2 (https://reactnavigation.org/blog/2018/05/07/react-navigation-2.0.html)
**react-navigation-redux-helpers** > ^2.0.9 (https://github.com/react-navigation/redux-helpers/releases/tag/2.0.0)
**react** > ^16.6.1
**react-test-renderer** > ^16.6.1
**react-native** > ^0.57.5 (https://github.com/react-native-community/releases/blob/master/CHANGELOG.md#0575) This version needs react on version `16.6.1`

The version 3 of `React-Navigation` doesn't work very well. The `StackNavigator` doesn't animate the page change smoothly), so I decided to go with `2.x.x`. 
Tested it on iOS and Android and added `-UseModernBuildSystem=NO` to `testIOS.sh`. (https://github.com/facebook/react-native/issues/20774)

**Comment by perry-mitchell** [comment]
Thanks @ph1p - We'll take a look at this tomorrow. Sorry for the delay!

**Comment by perry-mitchell** [comment]
@ph1p Is this still WIP or ready to go?

**Comment by ph1p** [comment]
Everything is all right (: If you want me to create a new branch, I can do that, because I experimented a little bit.

**Comment by ph1p** [comment]
> @ph1p Is this still WIP or ready to go?

I think it is ready to. Maybe later we can create a new repository just for translations and desktop, browser and mobile can implement it.

There was no way for me to run the tests. I also tried Bitrise and CircleCI (Android only).

**Comment by perry-mitchell** [comment]
Hi @ph1p - Once again, really sorry for the delay in moving this. I'll try to help get it through asap. Currently also trying to get Google Drive integration working, so perhaps this can go out with that!

**Comment by ph1p** [comment]
Hi @perry-mitchell ! No problem (: Take your time! I know it's not always easy to keep track of things. (Issues, PRs, different repositories and so on)

Please take look at https://github.com/buttercup/buttercup-mobile/pull/156#issuecomment-489357037 while reviewing.

**Comment by perry-mitchell** [comment]
Hi @ph1p - I've now gotten around to reviewing.. Really sorry for how long this has taken. I removed Google Drive from the deployment pathway as it was just too problematic to release just yet. Once the reviews are satisfied I can release this asap.

**Comment by perry-mitchell** [comment]
I also reverted the travis config.. this is what you'd submitted, for reference:

```
matrix:
  include:
    - language: node_js
      script: npm run test
      env:
        - NAME=Jest
      before_install:
        - npm i -g jest-cli
      install:
        - npm install
      node_js:
        - "8"
    - language: objective-c
      node_js:
        - "10.13.0"

      os: osx
      osx_image: xcode10.2

      install:
        - npm ci &>/dev/null
        - ./scripts/testIOS.sh "iPhone XS" "12.0" "build"
        - ./scripts/testIOS.sh "iPhone XS" "12.0" "test"

    - language: objective-c
      node_js:
        - "10.13.0"

      os: osx
      osx_image: xcode10.2

      install:
        - npm ci &>/dev/null
        - ./scripts/testIOS.sh "iPhone 6s" "10.3.1" "build"
        - ./scripts/testIOS.sh "iPhone 6s" "10.3.1" "test"

    - language: android
      sudo: required
      jdk: oraclejdk8
      env:
        - ANDROID_API=23
        - ADB_INSTALL_TIMEOUT=8 # minutes
        - ANDROID_ABI=x86_64
        - EMU_FLAVOR=default
        - ANDROID_HOME=/usr/local/android-sdk
        - TOOLS=${ANDROID_HOME}/tools
        - PATH=${ANDROID_HOME}:${ANDROID_HOME}/emulator:${TOOLS}:${TOOLS}/bin:${ANDROID_HOME}/platform-tools:${PATH}
        - ANDROID_NDK_HOME=${ANDROID_HOME}/ndk-bundle

      android:
        components:
          - tools

        licenses:
          - 'android-sdk-preview-license-.+'
          - 'android-sdk-license-.+'
          - 'google-gdk-license-.+'

      install:
        - echo 'count=0' > /home/travis/.android/repositories.cfg # Avoid harmless sdkmanager warning
        - echo y | sdkmanager "platform-tools" >/dev/null
        - echo y | sdkmanager "ndk-bundle"
        - echo y | sdkmanager "cmake;3.6.4111459"
        - echo y | sdkmanager "lldb;3.1"
        - echo y | sdkmanager "tools" >/dev/null # A second time per Travis docs, gets latest versions
        - echo y | sdkmanager "build-tools;28.0.3" >/dev/null # Implicit gradle dependency - gradle drives changes
        - echo y | sdkmanager "platforms;android-$ANDROID_API" >/dev/null # We need the ANDROID_API of the emulator we will run
        - echo y | sdkmanager "platforms;android-28" >/dev/null # We need the API of the current compileSdkVersion from gradle.properties
        - echo y | sdkmanager --channel=4 "emulator" # Experiment with canary, specifying 28.0.3 (prior version) did not work
        - echo y | sdkmanager "extras;android;m2repository" >/dev/null
        - echo y | sdkmanager "system-images;android-$ANDROID_API;$EMU_FLAVOR;$ANDROID_ABI" #>/dev/null # install our emulator
        - echo no | avdmanager create avd --force -n test -k "system-images;android-$ANDROID_API;$EMU_FLAVOR;$ANDROID_ABI" -c 10M
        - emulator -verbose -avd test -no-accel -no-snapshot -no-window $AUDIO -camera-back none -camera-front none -selinux permissive -qemu -m 2048 &
        - android-wait-for-emulator
        - adb shell input keyevent 82 &
        - curl -sL https://deb.nodesource.com/setup_8.x | sudo -E bash -
        - sudo apt-get install -y nodejs
        - npm install

      script:
        - ./scripts/testAndroid.sh
```

Would it be possible to ask you to create a new PR with it? I'd be happy to debug how it runs on travis with you on a separate PR, it was just too much to include in one.

**Comment by ph1p** [comment]
thanks for reviewing @perry-mitcheell (: I'll work on it

File: package.json [hunk 2]
```diff
@@ -15,7 +15,8 @@
     "test": "run-s test:format test:react-native",
     "test:android": "./scripts/testAndroid.sh",
     "test:format": "prettier-check '{{source,scripts,__tests__}/**/*.js,index.*.js,shim.js}'",
-    "test:react-native": "NODE_ENV=test jest --no-cache"
+    "test:react-native": "NODE_ENV=test jest --no-cache",
+    "i18next:parse-lang": "i18next-scanner --config ./config/i18next-scanner.config.js 'source/**/*.{js,jsx}'"
```
[the remark this answers is not in the document]
Comment (ph1p) [reply package.json]: sure

File: android/app/build.gradle [hunk 3]
```diff
@@ -152,22 +152,20 @@ android {
...
 dependencies {
+    implementation project(':react-native-localize')
     implementation project(':react-native-keychain')
     implementation project(':react-native-secure-storage')
     implementation project(':react-native-touch-id')
     implementation project(':react-native-randombytes')
     implementation project(':@yfuks/react-native-action-sheet')
     implementation project(':react-native-vector-icons')
-    implementation fileTree(dir: "libs", include: ["*.jar"])
-    implementation "com.android.support:appcompat-v7:23.0.1"
```
[the remark this answers is not in the document]
Comment (ph1p) [reply android/app/build.gradle]: It was a duplicate of line 169, so I removed it.
```
implementation "com.android.support:appcompat-v7:${rootProject.ext.supportLibVersion}"
```

[the remark this answers is not in the document]
Comment (perry-mitchell) [reply android/app/build.gradle]: Thanks!

**Comment by perry-mitchell** [comment]
@ph1p Let us know when you're ready for another review.

**Comment by ph1p** [comment]
I'm always ready.

**Comment by perry-mitchell** [comment]
Looks good to me. I can do some testing on this tomorrow. I'll leave it to @sallar  to make a review as well.

**Comment by perry-mitchell** [comment]
Apologies @ph1p once again for missing this - I'll try to handle the conflicts so you don't need to again. We upgrade react-navigation to get the OTP codes feature working, so I'll skip that in this MR.

**Comment by perry-mitchell** [comment]
Hi @ph1p - I've merged your branch with master over at https://github.com/buttercup/buttercup-mobile/pull/209 and it seems to work locally. There's now a few missing localisations as we added text in the OTP codes MR. There's also the error messages and some other pieces of text (eg. "OK") which need localising, but that should be quite easy I guess.

I did remove the changes for the react-navigation upgrade as we'd done the same, sorry, and had also removed the redux portion of it (as we were originally planning to do away with it anyway). I'm sorry if you spent any great amount of time on this.

Now the final part: What would you like to do? Should we take it from here and finish it off, or would you prefer to take the merge-commit into your branch and finalise it yourself? We're entirely happy either way.. just sorry that we left it sit for so long.

Thanks for the help and patience :)

**Comment by ph1p** [comment]
> Hi @ph1p - I've merged your branch with master over at #209 and it seems to work locally. There's now a few missing localisations as we added text in the OTP codes MR. There's also the error messages and some other pieces of text (eg. "OK") which need localising, but that should be quite easy I guess.
> 
> I did remove the changes for the react-navigation upgrade as we'd done the same, sorry, and had also removed the redux portion of it (as we were originally planning to do away with it anyway). I'm sorry if you spent any great amount of time on this.
> 
> Now the final part: What would you like to do? Should we take it from here and finish it off, or would you prefer to take the merge-commit into your branch and finalise it yourself? We're entirely happy either way.. just sorry that we left it sit for so long.
> 
> Thanks for the help and patience :)

Hi @perry-mitchell,
That was a long way, until now :D 

> I did remove the changes for the react-navigation upgrade as we'd done the same, sorry, and had also removed the redux portion of it (as we were originally planning to do away with it anyway). I'm sorry if you spent any great amount of time on this.

That's all right. It's your project and I see a PR more as a proposal and discussion basis to get the best out of it. And if just everything is merged without meaning, nobody is helped (:  

> Now the final part: What would you like to do? Should we take it from here and finish it off, or would you prefer to take the merge-commit into your branch and finalise it yourself? We're entirely happy either way.. just sorry that we left it sit for so long.

I can do it gladly. I want to help you and finish it in the next days ;) 
Should I create a new PR and we close this one?

**Comment by ph1p** [comment]
@perry-mitchell I think I'm done.
Let me know if something is missing or if I have to adjust the structure of the locale files.

File: package.json [hunk 4]
```diff
@@ -56,25 +58,27 @@
...
     "react-native-level-fs": "^3.0.0",
     "react-native-loading-spinner-overlay": "^0.5.2",
+    "react-native-localize": "^1.3.1",
     "react-native-navbar": "^1.7.1",
-    "react-native-os": "^1.2.5",
+    "react-native-os": "^1.2.6",
     "react-native-progress-wheel": "^1.0.4",
     "react-native-randombytes": "^3.5.3",
-    "react-native-reanimated": "^1.2.0",
+    "react-native-reanimated": "^1.4.0",
```
Comment (sallar) [inline package.json]: Does this compile? I'm pretty sure `react-native-reanimated` 1.4.0 doesnt work with RN < 0.60
https://github.com/software-mansion/react-native-reanimated/issues/488#issuecomment-559462197

File: scripts/testIOS.sh [hunk 5]
```diff
@@ -4,8 +4,9 @@ THISDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 SIMNAME=$1
 SIMVER=$2
+ACTION=$3
 
 echo Testing ${SIMNAME} with iOS ${SIMVER}...
 
 cd $THISDIR/../ios
-set -o pipefail && xcodebuild -scheme Buttercup -destination "platform=iOS Simulator,name=${SIMNAME},OS=${SIMVER}" build test | xcpretty -c
+set -o pipefail && xcodebuild -UseModernBuildSystem=NO -scheme Buttercup -destination "platform=iOS Simulator,name=${SIMNAME},OS=${SIMVER}" ${ACTION} | xcpretty -c
```
Comment (sallar) [inline scripts/testIOS.sh]: Is `UseModernBuildSystem=NO` because of the module I mentioned or is there another reason?

Comment (ph1p) [reply scripts/testIOS.sh]: I set the build system to legacy, because the default build system is not running :/

File: package.json (hunk 4 above)
Comment (ph1p) [reply package.json]: I'll take a look at it. Maybe it was because I was running `npm update`.

Comment (ph1p) [reply package.json]: I set it back to `1.2.0`.

File: scripts/testIOS.sh (hunk 5 above)
Comment (ph1p) [reply scripts/testIOS.sh]: I removed it

File: source/components/GroupsPage.js [hunk 6]
```diff
@@ -95,7 +97,11 @@ class GroupsPage extends Component {
                                             )
                                         }
                                     >
-                                        <Text>{group.title}</Text>
+                                        <Text>
+                                            {rawGroupIsTrash(group)
+                                                ? this.props.t("group.trash")
```
Comment (perry-mitchell) [inline source/components/GroupsPage.js]: I don't think we should do this. Titles should be pure and used entirely as the data within the vault - 1:1. Translating that is not a great idea. I would suggest that we translate this for when we **create**  these groups at vault creation.

File: source/components/GroupsPage.js [hunk 7]
```diff
@@ -46,7 +48,7 @@ class GroupsPage extends Component {
     static navigationOptions = ({ navigation }) => {
         const { params = {} } = navigation.state;
         const { groupID = "0", title, isTrash } = params;
-        const options = { title };
+        const options = { title: isTrash ? i18n.t("group.trash") : title };
```
Comment (perry-mitchell) [inline source/components/GroupsPage.js]: Same here - we shouldn't translate group/entry properties.

File: source/components/GroupsPage.js (hunk 6 above)
Comment (ph1p) [reply source/components/GroupsPage.js]: Ok. I removed both "Trash-Group" translations. It should best happen in this method, right? https://github.com/buttercup/buttercup-core/blob/7c60c6ed2c656790bbf23f42bfe20fc3942932a2/source/node/Archive.js#L295

**Comment by perry-mitchell** [comment]
Thanks @ph1p - Taking this forward to another branch so we can merge it with our rn upgrade and get it out! Great work on this!

**Pull request merged**